File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
private:
    bool isValid(int i, int j, vector<vector<char>>& grid){
        return i>=0 && j>=0 && i<grid.size() && j<grid[0].size();
    }
    
    int dx[4] = {0, 1, 0, -1};
    int dy[4] = {1, 0, -1, 0};
    
    void dfs(vector<vector<char>>& grid, int i, int j){
        
        for(int k=0;k<4;k++){
            int new_x = i + dx[k];
            int new_y = j + dy[k];
            
            if(isValid(new_x, new_y, grid) && grid[new_x][new_y] == '1'){
                grid[new_x][new_y] = '0';
                dfs(grid, new_x, new_y);
            }
        }
    }
public:
    int numIslands(vector<vector<char>>& grid) {
        int count=0;
        
        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid[0].size();j++){
                if(grid[i][j] == '1'){
                    grid[i][j] = '0';
                    dfs(grid, i, j);
                    count+=1;
                }
            }
        }
        
        return count;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.ts
const dx: number[] = [0, 1, 0, -1];
const dy: number[] = [1, 0, -1, 0];

function isValid(x: number, y:number, grid:string[][]): boolean{
    return x>=0 && y>=0 && x<grid.length && y<grid[0].length;
}


function dfs(x: number, y: number, grid:string[][]){
    for(let k=0;k<4;k++){
        let new_x = x + dx[k];
        let new_y = y + dy[k];
        if(isValid(new_x, new_y, grid) && grid[new_x][new_y] == "1"){
            grid[new_x][new_y] = "0";
            dfs(new_x, new_y, grid);
        }
    }
}


function numIslands(grid: string[][]): number {
    let ans=0;
    
    for(let i=0;i<grid.length;i++){
        for(let j=0;j<grid[0].length;j++){
            if(grid[i][j] == "1"){
                ans+=1;
                dfs(i, j, grid);
            }
        }
    }
    
    return ans;
};